internal/vm/vsock.go

```diff
@@ -15,125 +15,209 @@ package vm
+		}
+
+		err = tryConnRead(conn, vsockAckMsg(port), vsockAckMsgTimeout)
+		if err != nil {
+			return nil, vsockAckError{cause: err}
+		}
+		return conn, nil
+	}
+
+	tickerCh := time.NewTicker(vsockRetryInterval).C
 	var attemptCount int
-	for range time.NewTicker(retryInterval).C {
+	for {
 		attemptCount++
-		logger = logger.WithField("attempt", attemptCount)
+		logger := logger.WithField("attempt", attemptCount)
 
 		select {
 		case <-ctx.Done():
 			return nil, ctx.Err()
-		default:
-			conn, err := vsock.Dial(contextID, port)
-			if err == nil {
-				logger.WithField("connection", conn).Debug("vsock dial succeeded")
-				return conn, nil
-			}
-
-			// ENXIO and ECONNRESET can be returned while the VM+agent are still in the midst of booting
-			if isTemporaryNetErr(err) || isENXIO(err) || isECONNRESET(err) {
-				logger.WithError(err).Debug("temporary vsock dial failure")
+		case <-tickerCh:
+			conn, err := tryConnect()
+			if isTemporaryNetErr(err) {
+				err = errors.Wrap(err, "temporary vsock dial failure")
+				logger.WithError(err).Debug()
 				continue
+			} else if err != nil {
+				err = errors.Wrap(err, "non-temporary vsock dial failure")
+				logger.WithError(err).Error()
+				return nil, err
 			}
 
-			logger.WithError(err).Error("non-temporary vsock dial failure")
-			return nil, err
+			return conn, nil
 		}
 	}
-
-	panic("unreachable code") // appeases the compiler, which doesn't know the for loop is infinite
 }
 
-// VSockDialConnector provides an IOConnector interface to the VSockDial function.
-func VSockDialConnector(contextID, port uint32) IOConnector {
-	return func(procCtx context.Context, logger *logrus.Entry) <-chan IOConnectorResult {
-		returnCh := make(chan IOConnectorResult)
-
-		go func() {
-			defer close(returnCh)
-
-			conn, err := VSockDial(procCtx, logger, contextID, port)
-			returnCh <- IOConnectorResult{
-				ReadWriteCloser: conn,
-				Err:             err,
-			}
-		}()
-
-		return returnCh
-	}
+type vsockListener struct {
+	listener net.Listener
+	port     uint32
+	ctx      context.Context
+	logger   *logrus.Entry
 }
 
-func vsockAccept(reqCtx context.Context, logger *logrus.Entry, port uint32) (net.Conn, error) {
+// VSockListener returns a net.Listener implementation for guest-side Firecracker vsock
+// connections.
+func VSockListener(ctx context.Context, logger *logrus.Entry, port uint32) (net.Listener, error) {
 	listener, err := vsock.Listen(port)
 	if err != nil {
 		return nil, err
 	}
 
-	defer listener.Close()
+	return vsockListener{
+		listener: listener,
+		port:     port,
+		ctx:      ctx,
+		logger:   logger,
+	}, nil
+}
 
-	// Retries occur every 10ms up to vsockConnectTimeout
-	const retryInterval = 10 * time.Millisecond
-	ctx, cancel := context.WithTimeout(reqCtx, vsockConnectTimeout)
+func (l vsockListener) Accept() (net.Conn, error) {
+	ctx, cancel := context.WithTimeout(l.ctx, vsockRetryTimeout)
 	defer cancel()
 
+	tryAccept := func() (net.Conn, error) {
```

Comment:
Same question here: does this function need to be or benefit from being defined inline?